clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/AbstractEcsCachingAgent.java

```diff
@@ -0,0 +1,112 @@
+  protected AWSCredentialsProvider awsCredentialsProvider;
+  protected String region;
+  protected String accountName;
+  protected OnDemandMetricsSupport metricsSupport;
+
+  public AbstractEcsCachingAgent(String accountName, String region, AmazonClientProvider amazonClientProvider, AWSCredentialsProvider awsCredentialsProvider, Registry registry) {
+    this.accountName = accountName;
+    this.region = region;
+    this.amazonClientProvider = amazonClientProvider;
+    this.awsCredentialsProvider = awsCredentialsProvider;
+    this.metricsSupport = new OnDemandMetricsSupport(registry, this, EcsCloudProvider.ID + ":" + EcsCloudProvider.ID + ":${OnDemandAgent.OnDemandType.ServerGroup}");
+  }
+
+  protected abstract List<T> getItems(AmazonECS ecs, ProviderCache providerCache);
+
+  protected abstract CacheResult buildCacheResult(List<T> items);
+
+  @Override
+  public String getProviderName() {
+    return EcsProvider.NAME;
+  }
+
+  @Override
+  public String getOnDemandAgentType() {
+    return getAgentType();
+  }
+
+  @Override
+  public OnDemandMetricsSupport getMetricsSupport() {
+    return metricsSupport;
+  }
+
+  @Override
+  public Collection<Map> pendingOnDemandRequests(ProviderCache providerCache) {
+    return new LinkedList<>();
+  }
+
+  @Override
+  public boolean handles(OnDemandType type, String cloudProvider) {
+    return type.equals(OnDemandType.ServerGroup) && cloudProvider.equals(EcsCloudProvider.ID);
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    AmazonECS ecs = amazonClientProvider.getAmazonEcs(accountName, awsCredentialsProvider, region);
+    List<T> items = getItems(ecs, providerCache);
+    return buildCacheResult(items);
+  }
+
+  @Override
+  public OnDemandResult handle(ProviderCache providerCache, Map<String, ?> data) {
+    if (!data.get("account").equals(accountName) || !data.get("region").equals(region)) {
+      return null;
+    }
+
+    AmazonECS ecs = amazonClientProvider.getAmazonEcs(accountName, awsCredentialsProvider, region);
+
+    List<T> items = metricsSupport.readData(new Closure<List<T>>(this, this) {
+      public List<T> doCall() {
+        return getItems(ecs, providerCache);
+      }
+    });
+
+    storeOnDemand(providerCache, data);
+
+    CacheResult cacheResult = metricsSupport.transformData(new Closure<CacheResult>(this, this) {
+      public CacheResult doCall() {
+        return buildCacheResult(items);
+      }
+    });
+
+
+    Collection<String> typeStrings = new LinkedList<>();
+    for (AgentDataType agentDataType : getProvidedDataTypes()) {
+      typeStrings.add(agentDataType.toString());
+    }
+
+    OnDemandResult result = new OnDemandResult();
+    result.setAuthoritativeTypes(typeStrings);
+    result.setCacheResult(cacheResult);
+    result.setSourceAgentType(getAgentType());
+
+    return result;
+  }
+
+  protected void storeOnDemand(ProviderCache providerCache, Map<String, ?> data) {
+    //To be overwritten.
```

Comment:
Please add a TODO here

Reply:
Added:
`// TODO: Overwrite if needed.`

At the moment only `TaskCachingAgent` needs it.